File: src/config/systemPrompts.ts
import type { Agent } from '@/types';

const TEACHING_STYLE_INSTRUCTIONS: Record<string, string> = {
  strict: 'Be precise and thorough. Point out every error. Maintain high standards.',
  encouraging: 'Be warm and supportive. Celebrate progress. Gently correct errors.',
  casual: 'Be relaxed and friendly. Use natural language. Correct only major errors.',
  socratic: 'Ask guiding questions. Help the student discover answers on their own.',
};

const CORRECTION_MODE_INSTRUCTIONS: Record<string, string> = {
  immediate: 'Correct grammar errors immediately inline as they occur in the conversation.',
  endOfTurn: 'Respond naturally first, then summarize any corrections at the end of your turn.',
  onRequest: 'Only correct grammar when the student explicitly asks for correction.',
};

export function buildSystemPrompt(agent: Agent): string {
  const parts: string[] = [
    `You are ${agent.name}, a language learning partner.`,
    `Your role: ${agent.role}. Your personality: ${agent.personality}`,
    '',
    `The student's level is: ${agent.targetLevel}.`,
    `Teaching style: ${TEACHING_STYLE_INSTRUCTIONS[agent.style]}`,
    `Correction mode: ${CORRECTION_MODE_INSTRUCTIONS[agent.correctionMode]}`,
    '',
    'IMPORTANT RULES:',
    '1. Always respond naturally as your character.',
    '2. Keep responses concise (2-4 sentences for casual chat).',
    '3. Mix English with brief Chinese explanations when helpful.',
    '4. Adapt difficulty to the student\'s level.',
  ];

  if (agent.topics.length > 0) {
    parts.push(`5. Preferred conversation topics: ${agent.topics.join(', ')}.`);
  }

  parts.push(
    '',
    'When providing corrections, include a JSON block in this format:',
    '```correction',
    JSON.stringify({
      original: '<what the student said>',
      corrected: '<corrected version>',
      details: [
        {
          incorrect: '<error>',
          correct: '<fix>',
          explanation_zh: '<中文解释>',
          explanation_en: '<English explanation>',
          example: '<example sentence>',
        },
      ],
    }, null, 2),
    '```',
  );

  return parts.join('\n');
}

File: lang_buddy/lib/config/prompts/system_prompts.dart
import '../../domain/entities/agent.dart';

class SystemPromptBuilder {
  SystemPromptBuilder._();

  static String build(Agent agent) {
    return '''
$_basePrompt

## Your Identity
Name: ${agent.name} (${agent.nameZh})
Role: ${agent.role.label}
Personality: ${agent.personality}

## Teaching Configuration
Student Level: ${agent.targetLevel}
Correction Mode: ${agent.correctionMode.instruction}
Teaching Style: ${agent.style.instruction}

## Preferred Topics
${agent.topics.map((t) => '- $t').join('\n')}

$_correctionFormatInstruction
''';
  }

  static const _basePrompt = 'You are a language learning companion for a Chinese native speaker learning English.\n'
      'The student communicates via a chat app. Keep responses conversational and concise\n'
      '(under 150 words unless explaining grammar).\n'
      '\n'
      'IMPORTANT RULES:\n'
      '- When the student writes in Chinese, respond in Chinese but gently encourage switching to English.\n'
      '- When the student writes in English, ALWAYS respond in English.\n'
      '- When correcting errors, provide the explanation in Chinese so the student can understand, but show the correct English usage.\n'
      '- Use natural, conversational English — not textbook English.\n'
      '- Adapt your vocabulary complexity to the student\'s level.\n'
      '- Stay in character at all times.\n'
      '- Be encouraging and patient.';

  static final _correctionFormatInstruction =
      'When you need to correct grammar errors, include a correction block in your response using this exact format:\n'
      '\n'
      '${'`' * 3}correction\n'
      '{\n'
      '  "original": "the student\'s original text",\n'
      '  "corrected": "the corrected version",\n'
      '  "details": [\n'
      '    {\n'
      '      "incorrect": "specific error phrase",\n'
      '      "correct": "the corrected phrase",\n'
      '      "explanation_zh": "Chinese explanation",\n'
      '      "explanation_en": "English explanation of the correction",\n'
      '      "example": "An example sentence showing correct usage"\n'
      '    }\n'
      '  ]\n'
      '}\n'
      '${'`' * 3}\n'
      '\n'
      'Place the correction block at the end of your natural response.\n'
      'Only include corrections when there are actual grammar or usage errors.';
}
